staging/src/github.com/kubeedge/mapper-framework/_template/mapper/data/stream/img.go

```diff
@@ -0,0 +1,246 @@
+		return
+	}
+
+	outputFmtCtx.AvformatWriteHeader(nil)
+	ySize := width * height
+
+	// Write media data to media files
+	var packet avcodec.Packet
+	packet.AvNewPacket(ySize * 3)
+	defer packet.AvPacketUnref()
+	var gotPicture int
+	if cctx.AvcodecEncodeVideo2(&packet, frame, &gotPicture) < 0 {
+		klog.Error("Encode Error")
+		return
+	}
+	if gotPicture == 1 {
+		packet.SetStreamIndex(outStream.Index())
+		outputFmtCtx.AvWriteFrame(&packet)
+	}
+
+	outputFmtCtx.AvWriteTrailer()
+	if outputFmtCtx.Oformat().GetFlags()&avformat.AVFMT_NOFILE == 0 {
+		if err = outputFmtCtx.Pb().Close(); err != nil {
+			klog.Errorf("close output fmt context failed: %v", err)
+		}
+	}
+}
+
+// SaveFrame save frame.
+func SaveFrame(input string, outDir string, format string, frameCount int, frameInterval int) error {
+	// Open video file
+	avformat.AvDictSet(&avformat.Dict, "rtsp_transport", "tcp", 0)
+	avformat.AvDictSet(&avformat.Dict, "max_delay", "5000000", 0)
+
+	pFormatContext := avformat.AvformatAllocContext()
+	if avformat.AvformatOpenInput(&pFormatContext, input, nil, &avformat.Dict) != 0 {
+		return fmt.Errorf("Unable to open file %s", input)
+	}
+	// Retrieve stream information
+	if pFormatContext.AvformatFindStreamInfo(nil) < 0 {
+		return fmt.Errorf("Couldn't find stream information")
+	}
+	// Dump information about file onto standard error
+	pFormatContext.AvDumpFormat(0, input, 0)
+	// Find the first video stream
+	streamIndex := -1
+	for i := 0; i < int(pFormatContext.NbStreams()); i++ {
+		if pFormatContext.Streams()[i].CodecParameters().AvCodecGetType() == avformat.AVMEDIA_TYPE_VIDEO {
+			streamIndex = i
+			break
+		}
+	}
+	if streamIndex == -1 {
+		return fmt.Errorf("couldn't find video stream")
+	}
+	// Get a pointer to the codec context for the video stream
+	pCodecCtxOrig := pFormatContext.Streams()[streamIndex].Codec()
+	// Find the decoder for the video stream
+	pCodec := avcodec.AvcodecFindDecoder(pCodecCtxOrig.CodecId())
+	if pCodec == nil {
+		return fmt.Errorf("unsupported codec")
+	}
+	// Copy context
+	pCodecCtx := pCodec.AvcodecAllocContext3()
+	if pCodecCtx.AvcodecCopyContext((*avcodec.Context)(unsafe.Pointer(pCodecCtxOrig))) != 0 {
+		return fmt.Errorf("couldn't copy codec context")
+	}
+
+	// Open codec
+	if pCodecCtx.AvcodecOpen2(pCodec, nil) < 0 {
+		return fmt.Errorf("could not open codec")
+	}
+
+	// Allocate video frame
+	pFrame := avutil.AvFrameAlloc()
+
+	// Allocate an AVFrame structure
+	pFrameRGB := avutil.AvFrameAlloc()
+	if pFrameRGB == nil {
+		return fmt.Errorf("unable to allocate RGB Frame")
+	}
+	// Determine required buffer size and allocate buffer
+	numBytes := uintptr(avcodec.AvpictureGetSize(avcodec.AV_PIX_FMT_YUVJ420P, pCodecCtx.Width(),
+		pCodecCtx.Height()))
+	buffer := avutil.AvMalloc(numBytes)
+
+	// Assign appropriate parts of buffer to image planes in pFrameRGB
+	// Note that pFrameRGB is an AVFrame, but AVFrame is a superset
+	// of AVPicture
+	avp := (*avcodec.Picture)(unsafe.Pointer(pFrameRGB))
```

Comment:
Should we consider the case where type is invalid?

Reply:
I don't quite understand your question. Is there a problem with converting *avutil.Frame to *avcodec.Picture? The official documentation here shows that *avcodec.Picture is a superset of *avutil.Frame. You can refer to [this](https://github.com/giorgisio/goav/blob/master/example/tutorial01.go#L128)